chain/client/src/chunk_executor_actor.rs

```diff
@@ -95,92 +104,195 @@ pub struct ExecutorBlock {
             return;
-        };
-        for next_block_hash in next_block_hashes.clone() {
-            if let Err(err) = self.try_apply_chunks(&next_block_hash) {
-                tracing::error!(target: "chunk_executor", ?err, ?block_hash, "failed to apply chunk for block hash");
-            };
+        }
+        if let Err(err) = self.try_process_next_blocks(&block_hash) {
+            tracing::error!(target: "chunk_executor", ?err, ?block_hash, "failed to process next blocks");
         }
     }
 }
 
 impl Handler<ExecutorBlock> for ChunkExecutorActor {
     fn handle(&mut self, ExecutorBlock { block_hash }: ExecutorBlock) {
-        // We may have received receipts before the corresponding block.
-        let block = match self.chain_store.get_block(&block_hash) {
-            Ok(block) => block,
+        match self.try_apply_chunks(&block_hash) {
+            Ok(TryApplyChunksOutcome::Scheduled) => {}
+            Ok(TryApplyChunksOutcome::NotReady) => {
+                if let Err(err) = self.add_pending_block(block_hash) {
+                    tracing::error!(target: "chunk_executor", ?err, ?block_hash, "failed to add pending block");
+                }
+            }
+            Ok(TryApplyChunksOutcome::BlockAlreadyAccepted) => {
+                tracing::warn!(
+                    target: "chunk_executor",
+                    ?block_hash,
+                    "not expected to receive already executed block"
+                );
+            }
             Err(err) => {
-                tracing::error!(target: "chunk_executor", %block_hash, ?err, "failed to get block");
-                return;
+                tracing::error!(target: "chunk_executor", ?err, ?block_hash, "failed to apply chunk for block hash");
             }
-        };
-        let header = block.header();
-        let prev_block_hash = header.prev_hash();
-
-        self.next_block_hashes.get_or_insert_mut(*prev_block_hash, || Vec::new()).push(block_hash);
+        }
+    }
+}
 
-        if let Err(err) = self.try_apply_chunks(&block_hash) {
-            tracing::error!(target: "chunk_executor", ?err, ?block_hash, "failed to apply chunk for block hash");
+impl Handler<ExecutorApplyChunksDone> for ChunkExecutorActor {
+    fn handle(
+        &mut self,
+        ExecutorApplyChunksDone { block_hash, apply_results }: ExecutorApplyChunksDone,
+    ) {
+        if let Err(err) = self.process_apply_chunk_results(block_hash, apply_results) {
+            tracing::error!(target: "chunk_executor", ?err, ?block_hash, "failed to process apply chunk results");
+            return;
         };
+        assert!(self.blocks_in_execution.remove(&block_hash));
+        if let Err(err) = self.try_process_next_blocks(&block_hash) {
+            tracing::error!(target: "chunk_executor", ?err, ?block_hash, "failed to process next blocks");
+        }
     }
 }
 
+enum TryApplyChunksOutcome {
+    Scheduled,
+    NotReady,
+    BlockAlreadyAccepted,
```

Comment:
```suggestion
    BlockAlreadyExecuted,
```

optional nit: makes naming slightly more consistent